File: cpp/src/bitmap/ColorRef.cpp
#include "libreshockwave/bitmap/ColorRef.hpp"

#include <algorithm>
#include <stdexcept>

namespace libreshockwave::bitmap {
namespace {

int clampByte(int value) {
    return std::clamp(value, 0, 255);
}

} // namespace

ColorRef::Rgb::Rgb(int r, int g, int b)
    : r(clampByte(r)), g(clampByte(g)), b(clampByte(b)) {}

ColorRef::Rgb ColorRef::Rgb::fromPacked(std::uint32_t rgb) {
    return Rgb(static_cast<int>((rgb >> 16) & 0xFF),
               static_cast<int>((rgb >> 8) & 0xFF),
               static_cast<int>(rgb & 0xFF));
}

ColorRef::Rgb ColorRef::Rgb::fromHex(std::string hex) {
    if (!hex.empty() && hex.front() == '#') {
        hex.erase(hex.begin());
    }
    if (hex.empty() || hex.size() > 6) {
        throw std::invalid_argument("Invalid RGB hex color");
    }
    return fromPacked(static_cast<std::uint32_t>(std::stoul(hex, nullptr, 16)));
}

std::uint32_t ColorRef::Rgb::toPacked() const {
    return (static_cast<std::uint32_t>(r) << 16) |
           (static_cast<std::uint32_t>(g) << 8) |
           static_cast<std::uint32_t>(b);
}

std::uint32_t ColorRef::Rgb::toArgb() const {
    return 0xFF000000U | toPacked();
}

ColorRef::PaletteIndex::PaletteIndex(int index)
    : index(clampByte(index)) {}

ColorRef::Rgb ColorRef::PaletteIndex::resolve(const Palette* palette) const {
    if (palette == nullptr) {
        palette = &Palette::systemMacPalette();
    }
    const auto rgb = palette->getRGB(index);
    return Rgb(rgb[0], rgb[1], rgb[2]);
}

ColorRef::ColorRef(Rgb rgb) : value_(rgb) {}

ColorRef::ColorRef(PaletteIndex index) : value_(index) {}

ColorRef::Rgb ColorRef::toRgb(const Palette* palette) const {
    if (const auto* rgb = std::get_if<Rgb>(&value_)) {
        return *rgb;
    }
    return std::get<PaletteIndex>(value_).resolve(palette);
}

int ColorRef::toNearestPaletteIndex(const Palette* palette) const {
    if (const auto* index = std::get_if<PaletteIndex>(&value_)) {
        return index->index;
    }
    if (palette == nullptr) {
        palette = &Palette::systemMacPalette();
    }
    return palette->nearestIndex(std::get<Rgb>(value_).toPacked());
}

bool ColorRef::isRgb() const {
    return std::holds_alternative<Rgb>(value_);
}

bool ColorRef::isPaletteIndex() const {
    return std::holds_alternative<PaletteIndex>(value_);
}

ColorRef::Rgb ColorRef::black() {
    return Rgb(0, 0, 0);
}

ColorRef::Rgb ColorRef::white() {
    return Rgb(255, 255, 255);
}

} // namespace libreshockwave::bitmap

File: cpp/apps/tools/ts/export-template/src/runtime/ColorRef.ts
// Color reference: either a direct RGB value or a palette index. Port of
// cpp/include/libreshockwave/bitmap/ColorRef.hpp and cpp/src/bitmap/ColorRef.cpp.
//
// Note: the palette-index resolution path falls back to the system Mac palette when no palette is
// supplied (matching the C++ runtime). That fallback depends on the built-in palette tables, which
// are ported in Stage 3; until then, palette-index ColorRefs without an explicit palette throw.

import { Palette } from "./Palette.js";

function clampByte(value: number): number {
  return Math.max(0, Math.min(255, value | 0));
}

export class Rgb {
  readonly r: number;
  readonly g: number;
  readonly b: number;

  constructor(r: number, g: number, b: number) {
    this.r = clampByte(r);
    this.g = clampByte(g);
    this.b = clampByte(b);
  }

  static fromPacked(rgb: number): Rgb {
    return new Rgb((rgb >>> 16) & 0xff, (rgb >>> 8) & 0xff, rgb & 0xff);
  }

  static fromHex(hex: string): Rgb {
    let value = hex;
    if (value.startsWith("#")) {
      value = value.slice(1);
    }
    if (value.length === 0 || value.length > 6 || !/^[0-9a-fA-F]+$/.test(value)) {
      throw new Error("Invalid RGB hex color");
    }
    return Rgb.fromPacked(Number.parseInt(value, 16));
  }

  toPacked(): number {
    return ((this.r << 16) | (this.g << 8) | this.b) >>> 0;
  }

  toArgb(): number {
    return (0xff000000 | this.toPacked()) >>> 0;
  }

  static black(): Rgb {
    return new Rgb(0, 0, 0);
  }

  static white(): Rgb {
    return new Rgb(255, 255, 255);
  }
}

export class PaletteIndex {
  readonly index: number;

  constructor(index: number) {
    this.index = clampByte(index);
  }

  resolve(palette?: Palette | null): Rgb {
    const resolved = palette ?? null;
    if (resolved === null) {
      // C++ falls back to Palette::systemMacPalette(); ported in Stage 3.
      throw new Error("PaletteIndex.resolve without a palette requires the system Mac palette (Stage 3)");
    }
    const [r, g, b] = resolved.getRGB(this.index);
    return new Rgb(r, g, b);
  }
}

type ColorRefValue =
  | { tag: "rgb"; value: Rgb }
  | { tag: "index"; value: PaletteIndex };

export class ColorRef {
  private readonly value_: ColorRefValue;

  constructor(rgb: Rgb);
  constructor(index: PaletteIndex);
  constructor(arg: Rgb | PaletteIndex) {
    if (arg instanceof Rgb) {
      this.value_ = { tag: "rgb", value: arg };
    } else {
      this.value_ = { tag: "index", value: arg };
    }
  }

  toRgb(palette?: Palette | null): Rgb {
    if (this.value_.tag === "rgb") {
      return this.value_.value;
    }
    return this.value_.value.resolve(palette);
  }

  toNearestPaletteIndex(palette?: Palette | null): number {
    if (this.value_.tag === "index") {
      return this.value_.value.index;
    }
    const resolved = palette ?? null;
    if (resolved === null) {
      // C++ falls back to Palette::systemMacPalette(); ported in Stage 3.
      throw new Error("ColorRef.toNearestPaletteIndex without a palette requires the system Mac palette (Stage 3)");
    }
    return resolved.nearestIndex(this.value_.value.toPacked());
  }

  isRgb(): boolean {
    return this.value_.tag === "rgb";
  }

  isPaletteIndex(): boolean {
    return this.value_.tag === "index";
  }
}